### ISSUE TITLE
Code Cleanup

### ISSUE THREAD

**Issue by ianlivingstone** [mortendahl/tf-encrypted#262]
We had mentioned during our planning session on Monday that there were some abstractions and other code components we'd like to clean up.

Let's use this issue as a gathering place to bring together all of the different issues we've opened and discuss which we should attack and when.

**Sub Issues**

The following is a list of sub-issues that we've created along the way related to code clean up and refactoring of the API.

- Refactoring Tensors and API Cleanup - #261

**Comment by ianlivingstone** [comment]
@mortendahl @jvmancuso @bendecoste @justin1121 it'd be great if you could drop your general thoughts with respect to what we should tackle code and API clean up wise into this issue. We can then prioritize and begin tackling it.

**Comment by bendecoste** [comment]
# Backing tensor types / vs shared tensors

This doesn't bother me anymore but it drove me crazy when I was new to the code base.  We have `Pond<type>Tensor`s (`PondMasked`, `PondPrivate`, etc) which are implicitly shared between two people, as well as a plethora of different backing tensors `int100`, `int64`, `int32`, `oddImplicit`, etc.  Before I was up to speed on the code base I found it super hard to reason about what I was dealing with for any given block of code.  After a while I got used to it and it didn't bother me anymore, but I fear that this might be a big road block for new contributors.

I am not 100% clear on how we actually solve this because it ends up being a very useful abstraction.  Perhaps it is something that can simply be cleaned up via naming changes (e.g. `Int100Tensor` becomes `Int100BackingTensor`) or documentation.

# Amount of redirection for an op

When you perform an op on a PondTensor, there is quite a bit of redirection that happens before you get into the meat of what really happens behind the scenes.  For example, multiplying 2 pond tensors together

```
a * b -> __mul__ -> def mul -> prot.mul -> lift -> dispatch -> def _mul_type_type
```

(might not be exact I wrote that from memory)

Eventually you learn to just skip right to prot but again, this could possibly be improved.

# Writing an OP 3 times

The above example also illustrates a third frustration.  Writing a new OP involves writing the same boiler plate in a ton of different spots.  If you want to add an op, e.g. `mul` You need to declare

- the function on `PondTensor`
- the function on a ton of different backing types if you want to be complete (sometimes a backing tensor will just raise `NotImplementedError`
- the function on prot (this involves the top level function and a dispatch, so in this example you would need to write)
  - `def mul`
  - `def _mul_public_public`
  - `def _mul_public_private`
  - `def _mul_private_masked`
  - etc

This is a ton of functions to write a single function!

# Pond is getting huge!

Pond is 3300 lines of code and growing!  Would need to talk about a clean way to fix this up.  An obvious idea is to move dispatch functions into their own files. e.g.

```
pond.py
  - def mul
  - def sub

ops/mul.py
  - def _mul_private_private
  - def _mul_masked_masked
```

We can probably do even more but this strategy would remove ~1500 lines from Pond.

We could (and should) move PondTensors to their own files ~500 lines


# tl;dr

- backing types vs. top level tensors
- amount of redirection for an op
- amount of code you need to write to implement an op (same boilerplate each time)
- size of pond

**Comment by justin1121** [comment]
I agree mostly with what @bendecoste has said. One of the most important things to me is figuring out the best way to organize the different tensor types. I'm not sure the best way to do this but introducing a dtype like thing that tensorflow/numpy use might be interesting. I think the backing tensors would still need to have an underlying class related to each dtype though. Documentation would help greatly in guiding new contributors as well.

As for the redirection of an op, I'm not too sure how we can avoid this but it would be very interesting if we could find a way to clean this up!

+1000 for shrinking pond file size.

**Comment by bendecoste** [comment]
```
if isinstance(x, OddImplicitTensor) and isinstance(y, OddImplicitTensor):
        assert x.factory == y.factory, "Incompatible data types: {} and {}".format(x.factory, y.factory)
        return x, y

    if isinstance(x, OddImplicitTensor):

        if isinstance(y, int):
            return x, x.factory.tensor(np.array([y]))

        if isinstance(y, np.ndarray):
            return x, x.factory.tensor(y)

        if isinstance(y, PrimeTensor):
            return x, x.factory.tensor(y.value)

    if isinstance(y, OddImplicitTensor):

        if isinstance(x, int):
            return y.factory.tensor(np.array([x])), y

        if isinstance(x, np.ndarray):
            return y.factory.tensor(x), y

    raise TypeError("Don't know how to lift {} {}".format(type(x), type(y)))
```

**Comment by mortendahl** [comment]
@bendecoste do you remember where that code snippet is from?

**Comment by bendecoste** [comment]
> @bendecoste do you remember where that code snippet is from?

https://github.com/mortendahl/tf-encrypted/blob/master/tf_encrypted/tensor/odd_implicit.py#L175

It was an example of how all of the tensors need to implement this lift, if you added a new tensor type you would need to update all of the lift methods.

**Comment by mortendahl** [comment]
@bendecoste I think casting might allowing us to avoid this; problem is that we're mixing types, which we shouldn't really be doing in the first place (ints and floats being the exception)

**Comment by jvmancuso** [comment]
Wanted to make a note here that we've been exploring moving away from explicitly defined Protocols and moving the protocol layer down into the definition of each high-level tfe op.  This would mean something like:
```python
a = tfe.relu(x)  # defaults to exact with SecureNN
b = tfe.maxpool(a)  # defaults to using garbled circuits or SecureNN
with tfe.compose_protocol(protocol_config, ...):
    # each of the following ops is executed with whatever protocol is specified
    # in the protocol_config arg, potentially replacing the default protocol with another
    c = tfe.matmul(b)  # could be done with HE or n-party SPDZ depending on config
    d = tfe.relu(c)  # exact with securenn or SPDZ approximated depending on config
    e = tfe.maxpool(d)  # can override default GC with SecureNN depending on config
```
The details here are hazy, and this could use more work, but playing around with more flexible ways of composing hybrid protocols on the fly seems like an important piece of this conversation.

**Comment by mortendahl** [comment]
to @jvmancuso's comment, I imagine that each operation could essentially be represented by its own class:

- securenn/ExactReLU
- pond/ApproxReLU

with instances of these registered with the `tfe` module forwarding calls like `tfe.relu` to the instance currently selected by the configuration.

Somehow this registration should also take types into account, ie multiplication for fixed point values need to apply a truncation whereas multiplication for integers values don't. In other words, the map maintained by the registry is not just e.g `'mul'` -> `Mul` but perhaps `('mul', fixedpoint)` -> `FixedpointMul` and `('mul', integer)` -> `IntegerMul`.

FYI this line of thinking is heavily inspired by [Rust traits](https://doc.rust-lang.org/book/traits.html), which can be type parameterised, eg:
```rust
impl Add<u8> for u32 { ... }
impl Add<i8> for u32 { ... }
```
and would help compartmentalising protocols and make it clear where one protocol can be used and where not (for one tried to do `tfe` might throw an error when no operation has been registered for the types of the inputs).

**Comment by mortendahl** [comment]
In light of new knowledge after meeting the TensorFlow guys we should spend some more time understanding the TF core before moving in to a refactoring: maybe it'll change how we overload operations or it may give some ideas for registering secure ops (ie kernels in TF jargon)
